crates/ruff_linter/src/rules/ruff/rules/for_variable_used_after_block.rs

```diff
@@ -0,0 +1,109 @@
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_semantic::{BindingId, NodeId, Scope};
+use ruff_text_size::Ranged;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for variables defined in `for` loops that are used outside of their
+/// respective blocks.
+///
+/// ## Why is this bad?
+/// Usage of of a control variable outside of the block they're defined in will probably
+/// lead to flawed logic in a way that will likely cause bugs. The variable might not
+/// contain what you expect.
+///
+/// In Python, unlike many other languages, `for` loops don't define their own scopes.
+/// Therefore, usage of the control variables outside of the block will be the the value
+/// from the last iteration until re-assigned.
+///
+/// While this mistake is easy to spot in small examples, it can be hidden in larger
+/// blocks of code, where the the loop and downstream usage may not be visible at the
+/// same time.
+///
+/// ## Example
+/// ```python
+/// for x in range(10):
+///     pass
+///
+/// print(x)  # prints 9
+/// ```
+#[violation]
+pub struct ForVariableUsedAfterBlock {
+    control_var_name: String,
+}
+
+impl Violation for ForVariableUsedAfterBlock {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let ForVariableUsedAfterBlock { control_var_name } = self;
+
+        format!("`for` loop variable {control_var_name} is used outside of block")
+    }
+}
+
+/// Based on wemake-python-styleguide (WPS441) to forbid control variables after the block body.
+pub(crate) fn for_variable_used_after_block(
+    checker: &Checker,
+    scope: &Scope,
+    diagnostics: &mut Vec<Diagnostic>,
+) {
+    // Keep track of the node_ids of variable references that have already been
+    // accounted for as part of a different variable binding. This helps us avoid
+    // complaining when people use the same variable name in multiple blocks under the
+    // same scope.
+    let mut known_good_reference_node_ids: Vec<NodeId> = Vec::new();
+
+    let all_bindings: Vec<(&str, BindingId)> = scope.all_bindings().collect();
+    // We want to reverse the bindings so that we iterate in source order and shadowed
+    // bindings come first. This way we can gather `known_good_reference_node_ids` and
+    // mark off things as we go which will allow us to ignore later bindings trying to
+    // reference the same variable.
+    let reversed_bindings = all_bindings.iter().rev();
```

Comment:
I was able to remove the `rev` call and all tests keep passing. We should add a test that demonstrate why `rev` is necessary or remove the `rev` call (and the collecting of all bindings) if reversing in reversed-source-order is unnecessary.

I think if reverse iterating matters (because of redefinitions), then I'm not sure if only iterating over the `loop_var_bindings` is sufficient in the next step (what if the redeclaration is a non-loop variable)?

Reply:
Removed the reverse iteration over the bindings :white_check_mark: . This was only necessary to track `known_good_reference_node_ids` properly but [since that's not necessary anymore](https://github.com/astral-sh/ruff/pull/11769#discussion_r1710766228), can be removed.